### PR TITLE
Server manager support the NewUserConn operation

### PR DESCRIPTION
**Pull request by arugal**
### Issues associated with this PR

https://github.com/fatedier/frp/issues/1731

### CODE REVIEW

File: server/proxy/proxy.go [hunk 1]
```diff
@@ -41,6 +42,7 @@ type Proxy interface {
 	GetConf() config.ProxyConf
 	GetWorkConnFromPool(src, dst net.Addr) (workConn net.Conn, err error)
 	GetUsedPortsNum() int
+	GetPluginManager() *plugin.Manager
```
Comment (fatedier) [inline server/proxy/proxy.go:45]: Get pluginManager from `pxy.rc`.

File: models/plugin/server/types.go [hunk 2]
```diff
@@ -55,3 +57,17 @@ type NewWorkConnContent struct {
 	User UserInfo `json:"user"`
 	msg.NewWorkConn
 }
+
+type NewUserConn struct {
```
Comment (fatedier) [inline models/plugin/server/types.go:61]:
```suggestion
type NewUserConnContent struct {
```

File: models/plugin/server/manager.go [hunk 3]
```diff
@@ -189,3 +204,50 @@ func (m *Manager) NewWorkConn(content *NewWorkConnContent) (*NewWorkConnContent,
 	}
 	return content, nil
 }
+
+func (m *Manager) NewUserConn(userConn *NewUserConn) {
```
Comment (fatedier) [inline models/plugin/server/manager.go:208]: The logic of `NewUserConn` is same with `NewWorkConn`.

No need to use channel and cache.

File: server/proxy/proxy.go [hunk 4]
```diff
@@ -241,6 +248,9 @@ func HandleUserTcpConnection(pxy Proxy, userConn net.Conn, serverCfg config.Serv
 	xl.Debug("join connections, workConn(l[%s] r[%s]) userConn(l[%s] r[%s])", workConn.LocalAddr().String(),
 		workConn.RemoteAddr().String(), userConn.LocalAddr().String(), userConn.RemoteAddr().String())
 
+	// Record remote address for the access proxy
+	pxy.GetPluginManager().NewUserConn(plugin.CreateNewUserConn(pxy.GetName(), pxy.GetConf().GetBaseInfo().ProxyType, userConn.RemoteAddr()))
```
Comment (fatedier) [inline server/proxy/proxy.go:252]: Reference to `NewProxy` operation. We should handle plugin response to decide if we should reject the connection.

File: doc/server_plugin.md [hunk 5]
```diff
@@ -172,6 +172,22 @@ New work connection received from frpc (RPC sent after `run_id` is matched with
 }
 ```
 
+#### NewUserConn
+
+New user connection received from proxy (only support `tcp`, `http`, `https`) .
+
+The `NewUserConn` operation is triggered only on the first access to the proxy, clearing the cache every 24 hours.
```
[the remark this answers is not in the document]
Comment (arugal) [reply doc/server_plugin.md:179]: Got it, seems we understood it differently

**Comment by arugal** [comment]
Hi @fatedier I have made a preliminary modification and will submit a test feedback later.

File: server/proxy/proxy.go [hunk 6]
```diff
@@ -225,6 +232,23 @@ func HandleUserTcpConnection(pxy Proxy, userConn net.Conn, serverCfg config.Serv
...
 
+	// server plugin hook
+	rc := pxy.GetResourceController()
+	content := &plugin.NewUserConnContent{
+		User: pxy.GetUserInfo(),
+		NewUserConn: msg.NewUserConn{
+			ProxyName:  pxy.GetName(),
+			ProxyType:  pxy.GetConf().GetBaseInfo().ProxyType,
+			RemoteAddr: userConn.RemoteAddr().String(),
+			Timestamp:  time.Now().Unix(),
```
Comment (fatedier) [inline server/proxy/proxy.go:243]: We really need this timestamp?

File: server/proxy/proxy.go [hunk 7]
```diff
@@ -225,6 +232,23 @@ func HandleUserTcpConnection(pxy Proxy, userConn net.Conn, serverCfg config.Serv
...
+		},
+	}
+	_, err := rc.PluginManager.NewUserConn(content)
+	if err != nil {
+		xl.Warn("the user conn [%s] was rejected, err:%v", content.RemoteAddr, err)
```
Comment (fatedier) [inline server/proxy/proxy.go:248]: It's not a warning. Info is enough.

File: models/msg/msg.go [hunk 8]
```diff
@@ -128,6 +128,13 @@ type NewWorkConn struct {
 	Timestamp    int64  `json:"timestamp"`
 }
 
+type NewUserConn struct {
```
[the remark this answers is not in the document]
Comment (arugal) [reply models/msg/msg.go]: done

File: server/proxy/proxy.go (hunk 6 above)
Comment (arugal) [reply server/proxy/proxy.go]: done

File: server/proxy/proxy.go (hunk 7 above)
Comment (arugal) [reply server/proxy/proxy.go:244]: done

File: models/plugin/server/manager.go [hunk 9]
```diff
@@ -23,11 +23,14 @@ import (
 	"github.com/fatedier/frp/utils/xlog"
 )
 
+const newUserConnChanSize = 1000
```
[the remark this answers is not in the document]
Comment (arugal) [reply models/plugin/server/manager.go]: done

**Comment by fatedier** [comment]
Have you test it on all tcp, http and https type proxy mentioned in document?

**Comment by arugal** [comment]
> Have you test it on all tcp, http and https type proxy mentioned in document?

Doing

**Comment by arugal** [comment]
#### Configuration
+ frps
```
[common]
bind_port = 7000
vhost_http_port = 7001
vhost_https_port = t

[plugin.frp-notify]
addr = 127.0.0.1:80
path = /handler
ops = NewUserConn
```
+ frpc
```
[common]
server_addr = 127.0.0.1
server_port = 7000

[ssh]
type = tcp
local_ip = 127.0.0.1
local_port = 22
remote_port = 6000
```

### Result
```
NewUserConn - map[proxy_name:ssh proxy_type:tcp remote_addr:127.0.0.1:60720 user:map[metas:map[] run_id:104c9d5726c633ab user:]]
```
`http` and `https` are not currently supported, and I have changed the documentation.

**Comment by fatedier** [comment]
@arugal Proxy type `https, tcpmux, stcp, tcp` have all called the `HandleUserTcpConnection` function.

**Comment by arugal** [comment]
> @arugal Proxy type `https, tcpmux, stcp, tcp` have all called the `HandleUserTcpConnection` function.

The test results have been updated. Please review :)

**Comment by arugal** [comment]
[Server manager plugin](https://github.com/arugal/frp-notify/blob/17e2dff2c2bb4e3d1efdf6cdcbc1e0443e07b80f/pkg/server/manager.go#L83-L172) implementation.

**Pull request merged**